core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java

```diff
@@ -276,6 +277,30 @@ ClientRequestContext newDerivedContext(@Nullable HttpRequest req, @Nullable RpcR
      */
     void setResponseTimeout(Duration responseTimeout);
 
+    /**
+     * Returns {@link Response} timeout handler which is executed when
+     * the {@link Response} is not completely received within the allowed {@link #responseTimeoutMillis()}
+     * or the default {@link ClientOption#RESPONSE_TIMEOUT_MILLIS}.
+     */
+    @Nullable
+    Runnable responseTimeoutHandler();
+
+    /**
+     * Sets a handler to run when the response times out. {@code responseTimeoutHandler} must close the request,
+     * e.g., by calling {@link HttpRequestWriter#close(Throwable)}. If not set, the response will be closed with
+     * {@link ResponseTimeoutException}.
+     *
+     * <p>For example,
+     * <pre>{@code
+     *   HttpRequestWriter req = HttpRequest.streaming();
+     *   ctx.setResponseTimeoutHandler(() -> {
+     *      req.close(new IllegalStateException("Server is in a bad state."));
```

Comment:
Actually, I addressed [@anuraaga 's review](https://github.com/line/armeria/pull/2201#discussion_r336851759). If we change this, we should the documentation too. 
```java
{@code responseTimeoutHandler} must close the request, e.g., by calling {@link HttpRequestWriter#close(Throwable)}
```
WDYT? @anuraaga 🤔

Reply:
To be honest I'm not completely sure what the difference is between `HttpRequestWriter.close` and `HttpResponse.abort` for this handler - it's a point that confused me when trying to work on it.

But I figured since our example for `requestTimeoutHandler` in the server uses `HttpResponseWriter.close`, the client is supposed to be the opposite and call `HttpRequestWriter.close`

Reply:
The difference between `close` and `abort` is that `abort` works even if there is no subscriber. If there's no subscriber, `abort` prevents a new subscription and completes the completion future immediately.

Reply:
So when should a user use `close(Throwable)` vs `abort(Throwable)`? I think we need to be able to express this distinction clearly in Javadoc (probably of those methods themselves) for what seems to accomplish the same thing. Currently, we use `close(Throwable)` for all error handling, like when the connection is interrupted, and I'm not sure why not for timeout. Unless of course it makes sense to remove one of them, then it's simpler :D

Reply:
Don't think this was answered - can we make more clear when a user is supposed to use `abort` instead of `close`?

Currently, for server we call close here

https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcService.java#L211
https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L330

But is it supposed to use abort?

Reply:
I think `abort()` is generally better for handling timeout because it can clean up everything produced so far even if no subscription is made. However, it will not make a difference if you made a subscription before timeout occurs, so the benefit might be subtle.